src/call/mod.rs

```diff
@@ -34,3 +64,296 @@ impl RpcStatus {
+struct StreamingBase {
+    close_f: Option<BatchFuture>,
+    msg_f: Option<BatchFuture>,
+}
+
+impl StreamingBase {
+    fn new(close_f: Option<BatchFuture>) -> StreamingBase {
+        StreamingBase {
+            close_f: close_f,
+            msg_f: None,
+        }
+    }
+
+    fn poll(&mut self, call: &mut Call, skip_finish_check: bool) -> Poll<Option<Vec<u8>>, Error> {
+        if !skip_finish_check {
+            if let Some(ref mut close_f) = self.close_f {
+                match close_f.poll() {
+                    Ok(Async::Ready(_)) => return Ok(Async::Ready(None)),
+                    Err(e) => return Err(e),
+                    Ok(Async::NotReady) => {}
+                }
+            }
+        }
+
+        let mut bytes = None;
+        if let Some(ref mut msg_f) = self.msg_f {
+            bytes = try_ready!(msg_f.poll());
+            if bytes.is_none() {
+                return Ok(Async::Ready(None));
+            }
+        }
+
+        // so msg_f must be either stale or not initialised yet.
+        self.msg_f.take();
+        let msg_f = try!(call.start_recv_message());
+        self.msg_f = Some(msg_f);
+        if bytes.is_none() {
+            self.poll(call, true)
+        } else {
+            Ok(Async::Ready(bytes))
+        }
+    }
+}
+
+/// A helper struct for constructing Sink object for batch requests.
+struct SinkBase {
+    write_f: Option<BatchFuture>,
+    close_f: Option<BatchFuture>,
+    buf: Vec<u8>,
+    flags: u32,
+    send_metadata: bool,
+}
+
+impl SinkBase {
+    fn new(flags: u32, send_metadata: bool) -> SinkBase {
+        SinkBase {
+            write_f: None,
+            close_f: None,
+            buf: Vec::new(),
+            send_metadata: send_metadata,
+            flags: flags,
+        }
+    }
+
+    fn start_send<F, E>(&mut self, call: &mut Call, fill_buf: F) -> Result<bool>
+        where F: FnOnce(&mut Vec<u8>) -> result::Result<(), E>,
+              E: Into<Error>
+    {
+        if self.write_f.is_some() {
+            // try its best not to return false.
+            try!(self.poll_complete());
+            if self.write_f.is_some() {
+                return Ok(false);
+            }
+        }
+
+        self.buf.clear();
+        if let Err(e) = fill_buf(&mut self.buf) {
+            return Err(e.into());
+        }
+        let write_f = try!(call.start_send_message(&self.buf, self.flags, self.send_metadata));
+        self.write_f = Some(write_f);
+        self.send_metadata = false;
+        Ok(true)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Error> {
+        if let Some(ref mut write_f) = self.write_f {
+            try_ready!(write_f.poll());
+        }
+
+        self.write_f.take();
+        Ok(Async::Ready(()))
+    }
+
+    fn close(&mut self, call: &mut Call) -> Poll<(), Error> {
+        if self.close_f.is_none() {
+            if let Async::NotReady = try!(self.poll_complete()) {
```

Comment:
No. On the one hand, future requires buffered data should be flushed before actually closing:

> if a sink has buffered data, then it'll be flushed out during a close operation.

On the other hand, not flushing data may lead to wrong result for the streaming call.